### PR TITLE
Http cache tag aware store

### PR DESCRIPTION
**Pull request by bdunogier**
> Tag aware storage part of [EZP-22401](http://jira.ez.no/browse/EZP-22401)

Implements storage of tagged HTTP cache for the Symfony reverse proxy.

### TODO
- [ ] Rebase & fix ownership of commits
- [ ] Describe
- [ ] Finalize `doc/specifications/cache/multi_tagging.md`

### CODE REVIEW

**Comment by andrerom** [comment]
Should this go into feature branch like the others?

**Comment by bdunogier** [comment]
> Should this go into feature branch like the others?

aye !

File: doc/specifications/cache/multi_tagging.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+cache has been refreshed all other parallel requests are served stale cache item *(the old version)*.
+
+This is great for avoiding cache stamped effect, however for things like the Platform UI you'll need to take this into
+account. There are two ways of solving this:
+- Make UI code aware that operations won't update the response immediately, like in distributed systems with eventually consistency.
+- Setup UI on separate admin/backend domain where cache is shared with main domain, but where grace time is always 0.
+
+### Cache stamped protection
+
+Normally archived by having a percentage of randomness in expiry before the actual expiry time, so not several requests
```
Comment (andrerom) [inline doc/specifications/cache/multi_tagging.md]: This is a followup topic for persistence cache, Varnish does not really need it as long as we are able to use grace.

File: eZ/Publish/Core/MVC/Symfony/Cache/Http/ContentPurger.php [hunk 2]
```diff
@@ -10,14 +10,16 @@
...
- * indicating which locations to purge.
+ * When purging content by tags, purgeByRequest() would receive a Request object with xkey headers
+ * indicating which objects to purge.
  */
 interface ContentPurger extends RequestAwarePurger
 {
     /**
      * Purges all cached content.
      *
+     * @deprecated Use cache:clear, with multi tagging theoretically there shouldn't be need to delete all anymore from core.
```
Comment (bdunogier) [inline eZ/Publish/Core/MVC/Symfony/Cache/Http/ContentPurger.php]: Missing version

File: doc/specifications/cache/multi_tagging.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+end up trying to re generate a given cache, or several caches in parallel typically at the same time.
+
+### Avoiding to eager tag invalidation
```
Comment (andrerom) [inline doc/specifications/cache/multi_tagging.md]: Think we can skip for now, but was about trying to identify cases where we might invalidate too much cache still. Like for instance if we tag relations even if the are not displayed, or if we clear `content-type-<id>` for things not really affecting the content, like translating name of the content type.

File: eZ/Publish/Core/MVC/Symfony/Cache/Http/ContentPurger.php [hunk 4]
```diff
@@ -10,14 +10,16 @@
 
 /**
  * Interface allowing for HttpCache stores to purge specific content.
- * When purging content by locationId, purgeByRequest() would receive a Request object with X-Location-Id or X-Group-Location-Id headers
- * indicating which locations to purge.
+ * When purging content by tags, purgeByRequest() would receive a Request object with xkey headers
```
Comment (andrerom) [inline eZ/Publish/Core/MVC/Symfony/Cache/Http/ContentPurger.php]: yes (on purge)

**Comment by bdunogier** [comment]
Replaced by http://github.com/ezsystems/ezplatform-http-cache.

**Pull request closed without merging**